File: Samples/Beyond.NET.Sample.Android/app/src/androidTest/java/com/example/beyondnetsampleandroid/SystemConvertTests.kt
package com.example.beyondnetsampleandroid

import androidx.test.ext.junit.runners.AndroidJUnit4

import org.junit.Test
import org.junit.runner.RunWith

import org.junit.Assert.*

import com.example.beyondnetsampleandroid.dn.*
import java.util.Base64

@RunWith(AndroidJUnit4::class)
class SystemConvertTests {
    @Test
    fun testBooleanConversion() {
        val trueStringDN = "true".toDotNETString()
        val falseStringDN = "false".toDotNETString()
        val nonsenseStringDN = "nonsense".toDotNETString()

        val result1 = System_Convert.toBoolean(trueStringDN)
        assertTrue(result1)

        val result2 = System_Convert.toBoolean(falseStringDN)
        assertFalse(result2)

        assertThrows(DNException::class.java) { System_Convert.toBoolean(nonsenseStringDN) }
    }

    @Test
    fun testIntegerConversion() {
        val number1: Int = 123456789
        val number1StringDN = "$number1".toDotNETString()

        assertEquals(number1, System_Convert.toInt32(number1StringDN))

        val number2: Long = -123456789
        val number2StringDN = "$number2".toDotNETString()

        assertEquals(number2, System_Convert.toInt64(number2StringDN))

        val number3StringDN = "nonsense".toDotNETString()
        assertThrows(DNException::class.java) { System_Convert.toInt64(number3StringDN) }

        val number4StringDN = "nonsense".toDotNETString()
        assertThrows(DNException::class.java) { System_Convert.toUInt64(number4StringDN) }
    }

    @Test
    fun testBase64Conversion() {
        val text = "Hello World!"
        val textDN = text.toDotNETString()

        val utf8Encoding = System_Text_Encoding.uTF8_get()
        val textBytes = utf8Encoding.getBytes(textDN)
        val textAsBase64StringDN = System_Convert.toBase64String(textBytes)
        val textAsBase64String = textAsBase64StringDN.toKString()

        val textAsBase64ByteArray = Base64.getDecoder().decode(textAsBase64String)
        assertNotNull(textAsBase64ByteArray)

        val decodedText = String(textAsBase64ByteArray)
        assertNotNull(decodedText)

        assertEquals(text, decodedText)

        val textBytesRet = System_Convert.fromBase64String(textAsBase64StringDN)
        val textRet = utf8Encoding.getString(textBytesRet).toKString()
        assertEquals(text, textRet)
    }
}

File: Samples/Beyond.NET.Sample.Swift/Tests/System/SystemConvertTests.swift
import XCTest
import BeyondDotNETSampleKit

final class SystemConvertTests: XCTestCase {
    override class func setUp() {
        Self.sharedSetUp()
    }
    
    override class func tearDown() {
        Self.sharedTearDown()
    }
    
    func testBooleanConversion() throws {
        let trueStringDN = "true".dotNETString()
        let falseStringDN = "false".dotNETString()
        let nonsenseStringDN = "nonsense".dotNETString()
        
        let result1 = try System.Convert.toBoolean(trueStringDN)
        XCTAssertTrue(result1)
        
        let result2 = try System.Convert.toBoolean(falseStringDN)
        XCTAssertFalse(result2)
        
        do {
            let _ = try System_Convert.toBoolean(nonsenseStringDN)
            
            XCTFail("System.Convert.ToBoolean should throw")
        } catch { }
    }
    
    func testIntegerConversion() throws {
        let number1: Int32 = 123456789
        let number1StringDN = "\(number1)".dotNETString()
        
        XCTAssertEqual(number1,
                       try System.Convert.toInt32(number1StringDN))
        
        let number2: Int64 = -123456789
        let number2StringDN = "\(number2)".dotNETString()
        
        XCTAssertEqual(number2,
                       try System.Convert.toInt64(number2StringDN))
        
        let number3StringDN = "nonsense".dotNETString()
        XCTAssertThrowsError(try System.Convert.toInt64(number3StringDN))
        
        let number4StringDN = "nonsense".dotNETString()
        XCTAssertThrowsError(try System.Convert.toUInt64(number4StringDN))
    }
    
    func testBase64Conversion() throws {
        let text = "Hello World!"
        let textDN = text.dotNETString()
        
        let utf8Encoding = try System_Text_Encoding.uTF8
        let textBytes = try utf8Encoding.getBytes(textDN)
        let textAsBase64StringDN = try System_Convert.toBase64String(textBytes)
        let textAsBase64String = textAsBase64StringDN.string()
        
        guard let textAsBase64Data = textAsBase64String.data(using: .utf8) else {
            XCTFail("Failed to convert base64 string to data")
            
            return
        }
        
        guard let textData = Data(base64Encoded: textAsBase64Data) else {
            XCTFail("Failed to decode base64 encoded data")
            
            return
        }
        
        guard let decodedText = String(data: textData, encoding: .utf8) else {
            XCTFail("Failed to decode text data to string")
            
            return
        }
        
        XCTAssertEqual(text, decodedText)
        
        let textBytesRet = try System_Convert.fromBase64String(textAsBase64StringDN)
        let textRet = try utf8Encoding.getString(textBytesRet).string()
        XCTAssertEqual(text, textRet)
    }
}
